plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-2-field.tsx

```diff
@@ -0,0 +1,98 @@
+/**
+ * External dependencies
+ */
+import { ValidatedTextInput } from '@woocommerce/blocks-components';
+import { AddressFormValues, ContactFormValues } from '@woocommerce/settings';
+import {
+	useCallback,
+	useEffect,
+	useMemo,
+	useState,
+	Fragment,
+} from '@wordpress/element';
+import { objectHasProp } from '@woocommerce/types';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { AddressFieldProps } from './types';
+
+const Address2Field = < T extends AddressFormValues | ContactFormValues >( {
+	field,
+	fieldProps,
+	fieldsRef,
+	onChange,
+	values,
+}: AddressFieldProps< T > ): JSX.Element => {
+	const hasFieldValue = useMemo( () => {
+		return objectHasProp( values, 'address_2' ) && values.address_2 !== '';
+	}, [ values ] );
+
+	const [ hasFieldBeenModified, setHasFieldBeenModified ] = useState( false );
+
+	const isFieldRequired = field ? field.required : false;
+
+	const [ isFieldVisible, setFieldVisible ] = useState(
+		hasFieldValue || isFieldRequired
+	);
+
+	const toggleFieldVisibility = useCallback(
+		( event: React.MouseEvent< HTMLElement > ) => {
+			event.preventDefault();
+			setFieldVisible( ( prevVisibility: boolean ) => ! prevVisibility );
+		},
+		[]
+	);
+
+	useEffect( () => {
+		/**
+		 * Determine if the address 2 input field should be shown based on the following conditions:
+		 *
+		 * 1. If the address 2 field is required.
+		 * 2. If the address 2 field has a value.
+		 * 3. If the seller has edited the address 2 field.
+		 */
+		const shouldShowField =
+			isFieldRequired || hasFieldValue || hasFieldBeenModified;
+
+		setFieldVisible( shouldShowField );
+	}, [ isFieldRequired, hasFieldValue, hasFieldBeenModified ] );
+
+	return (
+		<Fragment key={ field.key }>
+			{ isFieldVisible ? (
+				<ValidatedTextInput
+					{ ...fieldProps }
+					key={ field.key }
+					ref={ ( el ) => ( fieldsRef.current[ field.key ] = el ) }
+					type={ field.type }
+					label={
+						isFieldRequired ? field.label : field.optionalLabel
+					}
+					className={ `wc-block-components-address-form__${ field.key }` }
+					value={ values[ field.key ] }
+					onChange={ ( newValue: string ) => {
+						onChange( {
+							...values,
+							[ field.key ]: newValue,
+						} );
+						setHasFieldBeenModified( true );
+					} }
+				/>
+			) : (
+				<button
+					key={ field.key }
+					className={
+						'wc-block-components-address-form__address_2-toggle'
+					}
+					onClick={ toggleFieldVisibility }
+				>
+					{ __( 'Add apartment, suite, etc.', 'woocommerce' ) }
```

Comment:
I still really think this should use the label, not doing now just means we will be asked to add a filter here in the future.

To give an example, the field can be renamed to something like, `box` and we would be asking for an apartment, suite in the link.

<img width="634" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/574ab18c-1a4a-4199-b1e5-7487e3a3f6e2">

<img width="638" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/b4e00a5f-ca27-46f1-8b63-1f525cb24fcc">
```suggestion
					{ sprintf(
						// translators: %s is the field label, example "Apartment, suite, etc."
						__( 'Add %s.', 'woocommerce' ),
						field.optionalLabel
					) }
```

I'd quickly double check with Nikki or the design if we want to keep the `( optional )` there or not.

Reply:
Thanks for clarifying this suggestion in Slack. I'm using `field.label` instead of `field.optionalLabel` to prevent that the link contains the suffix `(optional)`. This is aligned with @nikkivias's designs.